Export prebuilt initrd's and prebuilt signed uki's

AndrewAmmerlaan opened the pull request:
This changes the following:
- we add a default `dracut.conf`
- we add some extra dependencies required to create an useful initrd and unified kernel image, most importantly we add the cpu microcode packages required to update the microcode early in the initrd
- small fix in package.accept_keywords for a package rename
- new function `get_kernel_image_dir`, adapted from `dist-kernel-utils.eclass`, which we use to find the generatedinitrd and uki
- if we are building a kernel gpkg:
  - copy the generated initrd, then
  - switch out `uefi=no` with `uefi=yes` in our `dracut.conf`, then
  - reinstall the kernel again
  - copy the generated uki (it is signed with the same key as the main kernel image and kernel modules)

I'm far from an expert with docker, so there may be more elegant ways to do this. That being said, I can successfully boot my system with both the kernel+initrd and uki that I built in the docker container. I consider this experimental since I cannot be sure this will boot every system, that being said the pre-signed, pre-built uki is a requested feature and it is useful for those who want the added security of a verified initrd but do not want to generate and maintain their own key.

A next step would be to, by default, accept the key we use to sign everything here in `sys-boot/shim` MOK list. That is, if that is possible without building shim from source, I still have to investigate this.

Changes for the ebuild will follow in a separate PR soon. Note that no further changes are required to installkernel-gentoo or installkernel-systemd. For installkernel-gentoo we simply prevent the installation of the dracut plugin. And installkernel-systemd will find the initrd and uki.efi in the same directory as the kernel image and will use this instead of generating a new one.

CC @mgorny @thesamesam @ArsenArsen

AndrewAmmerlaan commented:
Fixed a permission error on the generated initrd/uki that would prevent `make rsync` from syncing the initrds/uki's.

mgorny commented on Dockerfile:
```diff
@@ -8,9 +8,24 @@ CMD cp -a /var/cache/binpkgs /tmp/binpkg \
  SECUREBOOT_SIGN_KEY="/tmp/signing_key.pem" \
  SECUREBOOT_SIGN_CERT="/tmp/signing_key.pem" emerge -1vB ${PKG} \
  && emerge -1vk ${PKG} \
+ && { [[ ${PKG} == *sys-kernel/gentoo-kernel* ]] \
```
Won't this cause the entire pipeline to fail if `[[ ... ]]` evaluates to false?

AndrewAmmerlaan replied:
Oops, yes it would do that, this needs a proper if/then. Fixing it now.

mgorny commented on Dockerfile:
```diff
@@ -8,9 +8,24 @@ CMD cp -a /var/cache/binpkgs /tmp/binpkg \
+   && gzip /tmp/initrds-and-ukis/$(basename ${kernel})-initramfs.img \
+   && chmod 666 /tmp/initrds-and-ukis/$(basename ${kernel})-initramfs.img.gz; \
+  done; } \
+  && { [[ ${UKI_SUPPORTED} ]] \
```
Likewise.

mgorny commented:
In general, I have serious doubts about this. We're talking about packing a lot of random userspace into the kernel -bin, and that's a major hassle. Old kernel versions may carry vulnerable dependencies for a long time (ok, people shouldn't be using these anyway but…). We'd also have to keep track of what goes into initramfs and update the `LICENSE` accordingly.

AndrewAmmerlaan commented:
> In general, I have serious doubts about this. We're talking about packing a lot of random userspace into the kernel -bin, and that's a major hassle. Old kernel versions may carry vulnerable dependencies for a long time (ok, people shouldn't be using these anyway but…). We'd also have to keep track of what goes into initramfs and update the `LICENSE` accordingly.

The LICENSE I hadn't thought about yet, but I'll open up the initrd later and check which files are included and where they come from. About this being a hassle and potentially fragile, I agree, that is why I added a big warning message to the ebuild in my other PR. Still there seems to be some interest in this, so I feel we can try this as an experimental feature and see if it works out. I also don't know (yet) how functional a generic initrd (and therefore UKI) image actually is, it might not work for all setups, but the only way to find out is to distribute a prebuilt UKI and ask folks to give it a try.

AndrewAmmerlaan commented:
This is now updated for systemd-255 and switched over to ukify for UKI generation. This has the nice additional benefit of not requiring the extra `emerge --config` step since now we get both an extractable initrd and extractable uki in one go.

Also fixed a problem where some random packages fail to build in the docker container due to interference from the `DEPS` environment variable.  And added more dependencies and dracut modules [recommended by upstream](https://github.com/dracutdevs/dracut/blob/4971f443726360216a4ef3ba8baea258a1cd0f3b/docs/README.generic) for a generic image.

I think this is now ready to go!

AndrewAmmerlaan commented:
I [backported](https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=32774e9d9e056831857e7aaf6aeed10498e9de39) a fix for the dracut pcrphase/pcrextend module with systemd-255, and re-added the module to our `dracutmodules` list.

AndrewAmmerlaan commented:
Small change today: Only pull in `sys-firmware/intel-microcode` on amd64 and x86, it makes no sense on arm64 or ppc64.

mgorny commented on build.bash:
```diff
@@ -213,6 +264,11 @@ export_vars() {
 	if [[ ${binpkg} ]]; then
+		if [[ ${binpkg} == kernel ]]; then
+			DOCKER_ARGS+=(
+				-v "${BINPKGROOT}/${target_arch}/initrds-and-ukis:/tmp/initrds-and-ukis"
```
To be honest, I don't like it at all that we're distributing the binpkg structure with stray files. That said, I suppose it would be better to keep them in the same directory as kernels, to make cleanup easier, i.e. `*6.6.1*` should remove both the kernel and related files.

I suppose this also means increasing mirror space usage a few times, doesn't it?

mgorny commented on os-release:
```diff
@@ -0,0 +1,8 @@
+NAME=Gentoo
```
Copy it from host system.

AndrewAmmerlaan replied:
In this case there is a reason to not copy if from the host. I changed `VERSION_ID` so the prebuilt UKI's are shown separately from those built manually in the bootloader menu.

If we want to use the host's version instead we can simply not specify `OSRelease` in `uki.conf`, the default is the host system's `/etc/os-release`, but I do like having them separated in the bootloader menu.

mgorny commented on uki.conf:
```diff
@@ -0,0 +1,16 @@
+SecureBootCertificate=/tmp/signing_key.pem
+SecureBootSigningTool=sbsign
+PCRPKey=/usr/src/linux/certs/signing_key.x509
+Splash=/tmp/gentoo-logo.bmp
```
Does this get compressed somehow or do we really add 1 MiB to the files?

AndrewAmmerlaan replied:
I don't think it gets compressed, but we can of course resize the source image (at the cost of quality). What kind of order of magnitude size should I be aiming for?

mgorny commented:
Ah, sorry, I've left the main commit and merged the other two fixes.

mgorny commented on build.bash:
```diff
@@ -26,30 +29,149 @@ export_vars() {
 			# we need systemd for initramfs
 			rc=systemd
+			# Optimize for smaller initrd
+			cflags='-mtune=generic -Oz -flto -pipe'
```
```suggestion
			cflags=${cflags/-O2/-Oz -flto}'
```

perhaps.
